Improve settings theme display

GuillaumeGomez opened the pull request:
This is a follow-up of #96958. In this PR, I make two UI changes:

 1. Themes are now handled in `<select>`. It's better because in case users add new themes, the UI will handle it better and it'll remain coherent. We can now potentially add as many themes as we want and will be handled much more easily.
 2. I added the brush icon alongside the theme settings' title.

It now looks like this:

![Screenshot from 2022-05-16 17-22-29](https://user-images.githubusercontent.com/3050060/168669192-7fcc2d66-8214-4760-8d25-4fee2767d223.png)

You can test it [here](https://rustdoc.crud.net/imperio/improve-settings-theme-display/doc/foo/index.html).

r? @jsha

rust-highfive commented:
Some changes occurred in HTML/CSS/JS.

cc @GuillaumeGomez,@Folyd,@jsha

GuillaumeGomez commented:
We were talking about it with @jsha. I think I'd be fine with radio buttons on their own line (I really don't like having them where they are located now). Also the brush image will be removed (we kinda all agree on this part hehe).

GuillaumeGomez commented:
It now looks like this:

![Screenshot from 2022-05-17 18-10-12](https://user-images.githubusercontent.com/3050060/168858464-6c10e52d-d9a8-4ab1-a140-84cd185e360e.png)

I also updated the [demo](https://rustdoc.crud.net/imperio/improve-settings-theme-display/doc/foo/index.html).

jsha commented on src/librustdoc/html/static/css/settings.css:
```diff
@@ -6,17 +6,10 @@
 	display: inline-block;
 	vertical-align: top;
-	font-size: 17px;
+	font-size: 1.06rem;
```
This is a nice change to use rem. But is there a good reason to have this font size slightly larger than the default? I think it would be fine at 1rem (i.e. don't override the default here).

GuillaumeGomez replied:
Done.

jsha commented on src/librustdoc/html/static/css/themes/dark.css:
```diff
@@ -3,6 +3,17 @@ body, #settings-menu #settings, #settings-menu #settings::before {
+}
+.setting-line .radio-line input:checked {
+	box-shadow: inset 0 0 0 3px #353535;
+	background-color: #2196f3;
```
What's this color? Can we reuse an existing color from the theme?

GuillaumeGomez replied:
It's the color from the "sliders" (the blue).

jsha replied:
I see that's set in settings.css:

```
input:checked + .slider {
    background-color: #2196F3;
}
```

But I think we always set colors in theme files, right? So I think that default blue color should get moved to light.css (even though it only applies to settings pages).

GuillaumeGomez replied:
Outch, that shouldn't have been. I'll move it into the theme files.

jsha commented on src/librustdoc/html/static/css/settings.css:
```diff
@@ -31,16 +24,33 @@
+	padding-bottom: 1px;
+}
+
+.radio-line .setting-name {
+	width: 100%;
```
Right now we have `.setting-line .radio-line` set to `flex` with `flex-wrap: wrap`, so the radio choices can wrap to the next line if the window is narrow. By making the heading 100% here you're making so that _always_ wraps, which will be confusing to the next person: "why put the flex there?"

So you should remove the `flex` and `flex-wrap` on `.setting-line .radio-line`. But keep it for `.setting-line .choices`. That will allow the choices to wrap if there are lots of them.

GuillaumeGomez replied:
Good point, removing them.

jsha reviewed (changes requested):
This looks nice! The larger radio buttons are a big improvement, and removing the borders around the labels is elegant.

Can you update the PR description with an up to date description and screenshot?

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-12`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/6474495479?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/d800cdcdadb1b0ccb082febd644095607cb8bebe/checks/6474495479/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
   Compiling rustdoc-themes v0.1.0 (/checkout/src/tools/rustdoc-themes)
    Finished release [optimized] target(s) in 0.54s
rustdoc: [check-theme] Starting tests! (Ignoring all other arguments)
 - Checking "/checkout/src/librustdoc/html/static/css/themes/dark.css"... FAILED
  Missing ".setting-line>.title" rule
Build completed unsuccessfully in 0:32:27
```

</details>

GuillaumeGomez commented:
Updated!

GuillaumeGomez commented:
I moved the "color" CSS rules into the theme files where they should always have been.

bors commented:
:umbrella: The latest upstream changes (presumably #97409) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

GuillaumeGomez commented:
Fixed merge conflicts.

jsha commented on src/librustdoc/html/static/css/settings.css:
```diff
@@ -6,41 +6,44 @@
 	display: inline-block;
 	vertical-align: top;
-	font-size: 17px;
+	font-size: 1rem;
```
You can remove this property entirely.

jsha commented on src/librustdoc/html/static/css/settings.css:
```diff
@@ -6,41 +6,44 @@
 	padding: 0.3em;
+	display: flex;
+	align-items: center;
+	cursor: pointer;
```
I think `pointer` is not needed for `.choice`. It is the default style for labels. It's only needed for the `input` ([for some reason radio buttons don't have pointer by default](https://ux.stackexchange.com/questions/23266/should-the-cursor-change-to-pointer-over-radio-buttons-checkboxes-their-labels)).

GuillaumeGomez replied:
No, apparently it's needed.

jsha reviewed (changes requested):
There's a regression with tab navigation: If you tab through the items in the settings menu, no outline appears when you have selected the radio buttons.

GuillaumeGomez commented:
Updated PR and the demo!

jsha commented:
The latest demo looks quite messed up (Chrome, Ubuntu, normal-width browser):

![image](https://user-images.githubusercontent.com/220205/171031535-51f69a69-8310-4b86-a87d-23cf28d34116.png)

GuillaumeGomez commented:
I cannot reproduce your bug. While looking for it, I saw another one for very small width screen so I fixed it (needed to add a `min-width` for `.slider`). I pushed and updated the demo as well.

jsha commented:
I can't reproduce the bug either. Looks good!

@bors r+ rollup

bors commented:
:pushpin: Commit 70db59ca3534d00227d903e10149e5fb677c0067 has been approved by `jsha`

<!-- @bors r=jsha 70db59ca3534d00227d903e10149e5fb677c0067 -->
<!-- homu: {"type":"Approved","sha":"70db59ca3534d00227d903e10149e5fb677c0067","approver":"jsha"} -->
